libzazen/ray.c

```diff
@@ -1,114 +1,169 @@
+                       ((fixed_float)local_coord_half_line.direction[1]).fixed,
+                       ((fixed_float)local_coord_half_line.direction[2]).fixed);
+  }
+#pragma GCC diagnostic pop
 }
 
-static void grab_ray_axis(struct zazen_ray_grab* grab,
-                          const struct timespec* time,
-                          struct zazen_ray_axis_event* event)
+static void zazen_ray_motion(struct zazen_ray* ray,
+                             struct zazen_cuboid_window* cuboid_window,
+                             struct zazen_ray_half_line local_coord_half_line,
+                             float min_distance)
 {
-  UNUSED(grab);
-  UNUSED(time);
-  UNUSED(event);
-  // TODO: send event to client
+  UNUSED(min_distance);
+  struct zazen_ray_client* ray_client;
+  struct wl_resource* ray_client_resource;
+  struct timeval tv;
+  uint32_t current_time_in_milles;
+
+  ray_client = zazen_ray_find_ray_client(
+      ray, wl_resource_get_client(cuboid_window->resource));
+  if (ray_client == NULL) return;
+
+  gettimeofday(&tv, NULL);
+  current_time_in_milles = tv.tv_sec * 1000 + tv.tv_usec / 1000;
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpedantic"
+  wl_list_for_each(ray_client_resource, &ray_client->ray_resources, link)
+  {
+    z11_ray_send_motion(
+        ray_client_resource, current_time_in_milles,
+        ((fixed_float)local_coord_half_line.origin[0]).fixed,
+        ((fixed_float)local_coord_half_line.origin[1]).fixed,
+        ((fixed_float)local_coord_half_line.origin[2]).fixed,
+        ((fixed_float)local_coord_half_line.direction[0]).fixed,
+        ((fixed_float)local_coord_half_line.direction[1]).fixed,
+        ((fixed_float)local_coord_half_line.direction[2]).fixed);
+  }
+#pragma GCC diagnostic pop
 }
 
-static void grab_ray_axis_source(struct zazen_ray_grab* grab,
-                                 enum wl_pointer_axis_source source)
+static void zazen_ray_leave(struct zazen_ray* ray,
+                            struct zazen_cuboid_window* cuboid_window,
+                            struct zazen_ray_half_line local_coord_half_line,
+                            float min_distance)
 {
-  UNUSED(grab);
-  UNUSED(source);
-  // TODO: send event to client
+  UNUSED(local_coord_half_line);
+  UNUSED(min_distance);
+
+  struct zazen_ray_client* ray_client;
+  struct wl_resource* ray_client_resource;
+  uint32_t serial;
+  zazen_cuboid_window_remove_highlight(cuboid_window);
+
+  ray_client = zazen_ray_find_ray_client(
+      ray, wl_resource_get_client(cuboid_window->resource));
+  if (ray_client == NULL) return;
+
+  serial = wl_display_next_serial(ray->seat->display);
+  wl_list_for_each(ray_client_resource, &ray_client->ray_resources, link)
+  {
+    z11_ray_send_leave(ray_client_resource, serial, cuboid_window->resource);
+  }
 }
 
-static void grab_ray_frame(struct zazen_ray_grab* grab)
+void zazen_ray_interact(struct zazen_ray* ray,
```

Comment:
綴ミス？interactという意図ならOK
```suggestion
void zazen_ray_intersect(struct zazen_ray* ray,
```